tools/repo-version/repo-version

```diff
@@ -0,0 +1,42 @@
+#! /usr/bin/env python3
+
+# Daemon BSD Source Code
+# Copyright (c) 2024, Daemon Developers
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of the <organization> nor the
+#    names of its contributors may be used to endorse or promote products
+#    derived from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from RepoVersion import RepoVersion
+
+import argparse
+
+parser = argparse.ArgumentParser(description="Print repository version string")
+
+parser.add_argument("-p", "--permissive", dest="is_permissive", help="ignore Git errors", action="store_true")
```

Comment:
Why would you want to ignore/silence Git errors? It would seem the script can't do anything useful if Git doesn't work.

Reply:
If Git doesn't work, and if the script is allowed to continue, it produces a version string using current date with same syntax (using `0` ans fallback tag and commit).

By using similar code in Urcheon I can generate versioned dpkdir of a work-in-progress folder even before committing the very first changes, and the fallback version string is compatible with what will be produced once there is a git repository, and once there is a git tag.

Example:

- `0-20240408-050944+dirty`: not a git repository yet, or git repository but no commit yet (date is world date).
- `0-20240408-051101-2bdab49`: git repository with committed data but not tag (date is latest commit date).
- `0-20240408-051101-2bdab49+dirty`: git repository with committed data but not tag, with uncommitted change (date is latest commit date).
- `0.54.1`: git repository, tag is current commit.
- `0.54.1+dirty`: git repository, tag is current commit, with uncommitted change.
- `0.54.1-20240408-051101-2bdab49`: git repository, untagged commits since latest tag (date is latest commit date).
- `0.54.1-20240408-051101-2bdab49+dirty`: git repository, untagged commits since latest tag, with uncommitted change (date is latest commit date).

Reply:
The script now tests if Git is available and working. The `--permissive` option can only continue on the fallback compute if Git is failing because the directory is not a git repository yet (the reason why there is a fallback).